src/internal/connector/onedrive/restore.go

```diff
@@ -696,26 +797,55 @@ func restoreData(
 		return "", details.ItemInfo{}, err
 	}
 
-	// Get a drive item writer
-	w, err := driveItemWriter(ctx, service, driveID, ptr.Val(newItem.GetId()), ss.Size())
-	if err != nil {
-		return "", details.ItemInfo{}, err
-	}
+	var (
+		w       io.Writer
+		written int64
+	)
 
-	iReader := itemData.ToReader()
-	progReader, closer := observe.ItemProgress(
-		ctx,
-		iReader,
-		observe.ItemRestoreMsg,
-		clues.Hide(name),
-		ss.Size())
+	// This is just to retry file upload, the uploadSession creation is not retried here
+	// We need extra logic to retry file upload as we have to pull the file again from kopia
+	for i := 0; i <= maxUploadRetries; i++ {
+		// Get a drive item writer
+		w, err = driveItemWriter(ctx, service, driveID, ptr.Val(newItem.GetId()), ss.Size())
+		if err != nil {
+			return "", details.ItemInfo{}, err
+		}
+
+		pname := name
+		iReader := itemData.ToReader()
+
+		if i > 0 {
+			pname = fmt.Sprintf("%s (retry %d)", name, i)
+
+			// If it is not the first try, we have to pull the file
+			// again from kopia. Ideally we could just seek the stream
+			// but we don't have a Seeker available here.
+			itemData, err := fetcher.Fetch(ctx, itemData.UUID())
+			if err != nil {
+				return "", details.ItemInfo{}, clues.Wrap(err, "getting data file")
+			}
+
+			iReader = itemData.ToReader()
+		}
+
+		progReader, closer := observe.ItemProgress(
+			ctx,
+			iReader,
+			observe.ItemRestoreMsg,
+			clues.Hide(pname),
+			ss.Size())
 
-	go closer()
+		go closer()
```

Comment:
Are we sure this will behave properly across loops?  Seems like you'd stack up multiple displays of the same item.

Reply:
Ahh, somehow missed that possibility. The `copyBufferWithStallCheck` was a result of me not understanding this was what was going on. I'll run a few more tests, but I think we don't need that function and a simple copy would do with the http lib failing in case of any issues.